File: 32970번： Autobus/Autobus.cpp
#include <iostream>
#include <sstream>
#include <string>
#include <vector>
#include <limits>
using namespace std;

struct Bus
{
    string from, to;
    int dep;
    int arr;
};

int toMinutes(const string &timeStr)
{
    int h, m;
    char colon;
    istringstream iss(timeStr);
    iss >> h >> colon >> m;
    return h * 60 + m;
}

int main()
{
    int n;
    cin >> n;

    vector<Bus> buses;

    string line;
    getline(cin, line);

    for (int i = 0; i < n; i++)
    {
        getline(cin, line);
        istringstream iss(line);
        string cities, times;
        iss >> cities >> times;

        size_t dash = cities.find('-');
        string from = cities.substr(0, dash);
        string to = cities.substr(dash + 1);

        size_t pos = times.find("--");
        string depStr = times.substr(0, pos);
        string arrStr = times.substr(pos + 2);

        int dep = toMinutes(depStr);
        int arr = toMinutes(arrStr);

        buses.push_back({from, to, dep, arr});
    }

    const int DAY = 1440;
    int best = numeric_limits<int>::max();

    for (const auto &bus1 : buses)
    {
        if (bus1.from == "Zagreb" && bus1.to == "Graz")
        {
            int effArr1 = bus1.arr;
            if (bus1.arr < bus1.dep)
            {
                effArr1 += DAY;
            }

            effArr1 += 1;

            for (const auto &bus2 : buses)
            {
                if (bus2.from == "Graz" && bus2.to == "Wroclaw")
                {
                    int duration2;
                    if (bus2.arr < bus2.dep)
                    {
                        duration2 = (bus2.arr + DAY - bus2.dep) + 1;
                    }
                    else
                    {
                        duration2 = (bus2.arr - bus2.dep) + 1;
                    }

                    int k = 0;
                    while (bus2.dep + k * DAY <= effArr1)
                    {
                        k++;
                    }

                    int effDep2 = bus2.dep + k * DAY;
                    int effArr2 = effDep2 + duration2;

                    int totalTime = effArr2 - bus1.dep;

                    if (totalTime < best)
                    {
                        best = totalTime;
                    }
                }
            }
        }
    }

    if (best == numeric_limits<int>::max())
    {
        cout << "NEMOGUCE" << endl;
    }
    else
    {
        int hours = best / 60;
        int minutes = best % 60;

        cout << hours << ":";
        if (minutes < 10)
            cout << "0";
        cout << minutes << endl;
    }

    return 0;
}

File: 32970번： Autobus/Autobus.py
def convert_to_minutes(time_str):
    h, m = map(int, time_str.split(':'))
    return h * 60 + m

def compute_effective_times(depart_str, arrive_str):
    dep = convert_to_minutes(depart_str)
    arr = convert_to_minutes(arrive_str)
    if arr < dep:
        arr += 1440

    return dep, arr + 1

import sys

def main():
    data = sys.stdin.read().strip().splitlines()
    if not data:
        return
    n = int(data[0].strip())

    trips_from_zagreb = []
    trips_from_graz = []

    for line in data[1:n+1]:
        parts = line.strip().split()
        route = parts[0]
        times = parts[1]

        depart_time, arrive_time = times.split('--')
        city_from, city_to = route.split('-')

        if city_from == "Zagreb" and city_to == "Graz":
            dep, eff_arr = compute_effective_times(depart_time, arrive_time)
            trips_from_zagreb.append((dep, eff_arr))
        elif city_from == "Graz" and city_to == "Wroclaw":
            base_dep, eff_arr = compute_effective_times(depart_time, arrive_time)
            trips_from_graz.append((base_dep, eff_arr))

    best_time = None

    for dep1, eff_arr1 in trips_from_zagreb:
        for base_dep2, base_eff_arr2 in trips_from_graz:
            if base_dep2 > eff_arr1:
                k = 0
            else:
                k = (eff_arr1 - base_dep2) // 1440 + 1

            abs_dep2 = base_dep2 + k * 1440

            if abs_dep2 <= eff_arr1:
                continue

            abs_eff_arr2 = base_eff_arr2 + k * 1440
            total_time = abs_eff_arr2 - dep1
            if best_time is None or total_time < best_time:
                best_time = total_time

    if best_time is None:
        print("NEMOGUCE")
    else:
        hours = best_time // 60
        minutes = best_time % 60
        print(f"{hours}:{minutes:02d}")

if __name__ == "__main__":
    main()